File: forms/neurodiversity-adjustment-response/front-end-with-svelte/src/lib/engine/outcome-rules.ts
import type { NeurodiversityAdjustmentResponse, OutcomeClassification, FiredRule } from './types';

/**
 * Axis A — outcome classification.
 *
 * Maps the employer's overall decision onto the outcome band. First match wins;
 * exactly one rule fires. Rule IDs are stable and identical across every
 * front-end and the back-end.
 */
export function classifyOutcome(r: NeurodiversityAdjustmentResponse): {
	outcomeClassification: OutcomeClassification;
	firedRules: FiredRule[];
} {
	const firedRules: FiredRule[] = [];

	switch (r.overallDecision) {
		case 'agreed':
			firedRules.push({
				ruleId: 'R-OUTCOME-AGREED',
				axis: 'outcome',
				category: 'fully-agreed',
				description: 'All requested adjustments agreed.'
			});
			return { outcomeClassification: 'fully-agreed', firedRules };
		case 'partially-agreed':
			firedRules.push({
				ruleId: 'R-OUTCOME-PARTIAL',
				axis: 'outcome',
				category: 'partially-agreed',
				description: 'Some requested adjustments agreed.'
			});
			return { outcomeClassification: 'partially-agreed', firedRules };
		case 'alternative-offered':
			firedRules.push({
				ruleId: 'R-OUTCOME-ALTERNATIVE',
				axis: 'outcome',
				category: 'alternative-offered',
				description: 'Alternative adjustments offered in place of those requested.'
			});
			return { outcomeClassification: 'alternative-offered', firedRules };
		case 'declined':
			firedRules.push({
				ruleId: 'R-OUTCOME-DECLINED',
				axis: 'outcome',
				category: 'declined',
				description: 'Requested adjustments declined.'
			});
			return { outcomeClassification: 'declined', firedRules };
		case 'deferred':
			firedRules.push({
				ruleId: 'R-OUTCOME-DEFERRED',
				axis: 'outcome',
				category: 'deferred',
				description: 'Decision deferred pending further information or assessment.'
			});
			return { outcomeClassification: 'deferred', firedRules };
		default:
			firedRules.push({
				ruleId: 'R-OUTCOME-UNSPECIFIED',
				axis: 'outcome',
				category: 'unspecified',
				description: 'No overall decision recorded yet.'
			});
			return { outcomeClassification: '', firedRules };
	}
}

File: forms/neurodiversity-adjustment-response/back-end-with-loco/src/engine/outcome_rules.rs
//! Axis A — outcome classification.
//!
//! Classifies the employer's overall decision. Exactly one rule fires (first
//! match wins):
//! - `fully-agreed`: all requested adjustments agreed.
//! - `partially-agreed`: some requested adjustments agreed.
//! - `alternative-offered`: alternatives offered in place of those requested.
//! - `declined`: requested adjustments declined.
//! - `deferred`: decision deferred pending further information.
//! - `''`: no overall decision recorded yet.
//!
//! Rule IDs are stable and identical across every front-end and the back-end.

use super::types::{FiredRule, NeurodiversityAdjustmentResponse, OutcomeClassification};

/// Classify the outcome, returning the classification plus the fired rule.
#[must_use]
pub fn classify_outcome(
    r: &NeurodiversityAdjustmentResponse,
) -> (OutcomeClassification, Vec<FiredRule>) {
    let (classification, rule_id, description) = match r.overall_decision.as_str() {
        "agreed" => (
            "fully-agreed",
            "R-OUTCOME-AGREED",
            "All requested adjustments agreed.",
        ),
        "partially-agreed" => (
            "partially-agreed",
            "R-OUTCOME-PARTIAL",
            "Some requested adjustments agreed.",
        ),
        "alternative-offered" => (
            "alternative-offered",
            "R-OUTCOME-ALTERNATIVE",
            "Alternative adjustments offered in place of those requested.",
        ),
        "declined" => (
            "declined",
            "R-OUTCOME-DECLINED",
            "Requested adjustments declined.",
        ),
        "deferred" => (
            "deferred",
            "R-OUTCOME-DEFERRED",
            "Decision deferred pending further information or assessment.",
        ),
        _ => (
            "",
            "R-OUTCOME-UNSPECIFIED",
            "No overall decision recorded yet.",
        ),
    };

    let fired_rules = vec![FiredRule {
        rule_id: rule_id.to_string(),
        axis: "outcome".to_string(),
        category: if classification.is_empty() {
            "unspecified".to_string()
        } else {
            classification.to_string()
        },
        description: description.to_string(),
    }];

    (classification.to_string(), fired_rules)
}
